win_etw_macros/src/lib.rs

```diff
@@ -616,23 +655,60 @@ fn trace_logging_events_core(attr: TokenStream, item_tokens: TokenStream) -> Tok
         impl #provider_ident {
             #provider_impl_items
         }
-
-        #[doc(hidden)]
-        #[allow(non_snake_case)]
-        mod #provider_mod_ident {
-            /*
-            pub(crate) mod event_descriptors {
-                #![allow(non_upper_case_globals)]
-                #( #event_descriptors )*
-            }
-            */
-        }
     });
 
     output.extend(errors.into_iter().map(|e| e.to_compile_error()));
     output
 }
 
+/// Creates a fragment of code (statements) which will register the
+/// provider traits for this provider.
+fn create_register_provider_traits(
+    provider_name: &str,
+    provider_group_guid: Option<&Uuid>,
+) -> TokenStream {
+    fn align2(v: &mut Vec<u8>) {
+        if v.len() % 2 != 0 {
+            v.push(0);
+        }
+    }
+
+    let mut traits_bytes: Vec<u8> = Vec::new();
+    traits_bytes.push(0); // reserve space for TraitsSize (u16)
+    traits_bytes.push(0);
+
+    traits_bytes.extend_from_slice(provider_name.as_bytes());
+    traits_bytes.push(0);
+    align2(&mut traits_bytes);
+
+    if let Some(provider_group_guid) = provider_group_guid {
+        // Add trait for provider guid
+        let provider_guid_trait_offset = traits_bytes.len();
+        traits_bytes.push(0); // reserve space for TraitSize (u16)
+        traits_bytes.push(0);
+        traits_bytes.push(ETW_PROVIDER_TRAIT_TYPE_GROUP);
+        traits_bytes.extend_from_slice(provider_group_guid.as_bytes());
+        let provider_guid_trait_len = traits_bytes.len() - provider_guid_trait_offset;
+        // Set TraitSize (u16)
+        traits_bytes[provider_guid_trait_offset] = provider_guid_trait_len as u8;
```

Comment:
Is this supposed to be big endian encoded?

Reply:
Host-encoded, so no conversion.